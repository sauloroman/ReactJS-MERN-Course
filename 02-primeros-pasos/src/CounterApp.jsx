import { useState } from "react"
import PropTypes from "prop-types"

export const CounterApp = ({ value }) => {

  const [counter, setCounter] = useState(value)

  const onIncreaseCounter = () => {
    setCounter( counter => counter + 1 )
  }

  const onDecreaseCounter = () => {
    if ( counter === 1 ) return
    setCounter( counter =>  counter - 1 )
  }

  const onResetCounter = () => setCounter(value)

  return (
    <div>
      <h1>{counter}</h1>
      
      <button onClick={ onDecreaseCounter }>-1</button>
      <button onClick={ onResetCounter}>Reset</button>
      <button onClick={ onIncreaseCounter }>+1</button>
    </div>
  )
}

CounterApp.propTypes = {
  value: PropTypes.number.isRequired
}