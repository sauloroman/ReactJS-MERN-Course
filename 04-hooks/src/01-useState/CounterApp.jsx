import { useState } from 'react';

export const CounterApp = () => {
  const [{ counter1, counter2, counter3 }, setCounter] = useState({
    counter1: 10,
    counter2: 20,
    counter3: 30,
  });

  return (
    <>
      <h1 className="heading">UseState &mdash; CounterApp</h1>

      <div className="counters">
        <div className="counter">
          <h2 className="counter__value" data-testid="counter-1">{counter1}</h2>
          <span className="counter__name">Contador 1</span>
          <button
            onClick={() =>
              setCounter((current) => ({
                ...current,
                counter1: counter1 + 1,
              }))
            }
            className="counter__btn"
          >
            Aumentar 1
          </button>
        </div>
        <div className="counter">
          <h2 className="counter__value" data-testid="counter-2">{counter2}</h2>
          <span className="counter__name">Contador 2</span>
          <button
            onClick={() =>
              setCounter((current) => ({
                ...current,
                counter2: counter2 + 1,
              }))
            }
            className="counter__btn"
          >
            Aumentar 2
          </button>
        </div>
        <div className="counter">
          <h2 className="counter__value" data-testid="counter-3">{counter3}</h2>
          <span className="counter__name">Contador 3</span>
          <button
            onClick={() =>
              setCounter((current) => ({
                ...current,
                counter3: counter3 + 1,
              }))
            }
            className="counter__btn"
          >
            Aumentar 3
          </button>
        </div>
      </div>
    </>
  );
};
