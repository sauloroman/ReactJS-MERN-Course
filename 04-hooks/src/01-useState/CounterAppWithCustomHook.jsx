import { useCounter } from '../hooks';

export const CounterAppWithCustomHook = () => {
  const { counter, onDecrementCounter, onIncrementCounter, onResetCounter } =
    useCounter(1);

  return (
    <>
      <h1 className="heading">UseState &mdash; CounterAppCustomHook</h1>

      <div className="counter counter--unique">
        <span data-testid="counter" className="counter__value">{counter}</span>
        <span className="counter__name">Contador</span>

        <div className="counter__buttons">
          <button onClick={() => onIncrementCounter()} className="counter__btn">+1</button>
          <button onClick={onResetCounter} className="counter__btn">Reset</button>
          <button onClick={() => onDecrementCounter()} className="counter__btn">-1</button>
        </div>
      </div>
    </>
  );
};
