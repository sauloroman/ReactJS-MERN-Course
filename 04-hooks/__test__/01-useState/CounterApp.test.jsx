import { render, fireEvent } from '@testing-library/react'
import { CounterApp } from '../../src/01-useState/CounterApp'

describe('Tests in CounterApp', () => {

  test('It must match with the snapshot', () => {
    const { container } = render(<CounterApp />)
    expect( container ).toMatchSnapshot() 
  })

  test('It must show the three counters correctly', () => {

    const { getAllByRole } = render(<CounterApp />)

    const counters = getAllByRole('heading', { level: 2 })

    expect( counters.length ).toBe(3)

    expect( counters[0].textContent ).toBe('10')
    expect( counters[1].textContent ).toBe('20')
    expect( counters[2].textContent ).toBe('30')

  })

  test('It must increment the appropiate counter depending the button pushed', () => {

    const { getByText, getByTestId } = render(<CounterApp />)

    const buttonCounter1 = getByText('Aumentar 1')
    fireEvent.click( buttonCounter1 )
    const counter1 = getByTestId('counter-1')

    const buttonCounter2 = getByText('Aumentar 2')
    fireEvent.click( buttonCounter2 )
    const counter2 = getByTestId('counter-2')

    expect( counter1.textContent ).toBe("11")
    expect( counter2.textContent ).toBe("21")

  })

})