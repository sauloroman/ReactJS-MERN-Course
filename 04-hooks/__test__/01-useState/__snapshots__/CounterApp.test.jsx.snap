// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in CounterApp It must match with the snapshot 1`] = `
<div>
  <h1
    class="heading"
  >
    UseState â€” CounterApp
  </h1>
  <div
    class="counters"
  >
    <div
      class="counter"
    >
      <h2
        class="counter__value"
        data-testid="counter-1"
      >
        10
      </h2>
      <span
        class="counter__name"
      >
        Contador 1
      </span>
      <button
        class="counter__btn"
      >
        Aumentar 1
      </button>
    </div>
    <div
      class="counter"
    >
      <h2
        class="counter__value"
        data-testid="counter-2"
      >
        20
      </h2>
      <span
        class="counter__name"
      >
        Contador 2
      </span>
      <button
        class="counter__btn"
      >
        Aumentar 2
      </button>
    </div>
    <div
      class="counter"
    >
      <h2
        class="counter__value"
        data-testid="counter-3"
      >
        30
      </h2>
      <span
        class="counter__name"
      >
        Contador 3
      </span>
      <button
        class="counter__btn"
      >
        Aumentar 3
      </button>
    </div>
  </div>
</div>
`;
